File: app/src/main/java/com/example/theloop/WidgetUpdateWorker.java
package com.example.theloop;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.ListenableWorker;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.example.theloop.models.WeatherResponse;
import com.example.theloop.network.RetrofitClient;
import com.example.theloop.network.WeatherApiService;
import com.example.theloop.utils.AppConstants;
import com.google.gson.Gson;

import java.io.IOException;

import retrofit2.Call;
import retrofit2.Response;

public class WidgetUpdateWorker extends Worker {

    private static final String TAG = "WidgetUpdateWorker";

    public WidgetUpdateWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        Log.d(TAG, "Fetching weather for widget update...");

        android.content.SharedPreferences prefs = getApplicationContext().getSharedPreferences(AppConstants.PREFS_NAME, Context.MODE_PRIVATE);
        double lat;
        double lon;

        String latStr = prefs.getString(AppConstants.KEY_LATITUDE, null);
        String lonStr = prefs.getString(AppConstants.KEY_LONGITUDE, null);

        if (latStr == null || lonStr == null) {
            Log.w(TAG, "No location available for widget update. Skipping weather fetch.");
            return Result.success();
        }

        try {
            lat = Double.parseDouble(latStr);
            lon = Double.parseDouble(lonStr);
        } catch (NumberFormatException e) {
            Log.w(TAG, "Could not parse lat/lon from SharedPreferences", e);
            return Result.failure();
        }

        // Fetch Weather
        try {
            String unit = prefs.getString(AppConstants.KEY_TEMP_UNIT, AppConstants.DEFAULT_TEMP_UNIT);

            WeatherApiService apiService = RetrofitClient.getClient().create(WeatherApiService.class);
            Call<WeatherResponse> call = apiService.getWeather(lat, lon,
                    "temperature_2m,weather_code", "weather_code,temperature_2m_max,temperature_2m_min", unit, "auto");

            Response<WeatherResponse> response = call.execute(); // Synchronous
            if (response.isSuccessful() && response.body() != null) {
                // Save to cache
                String json = new Gson().toJson(response.body());
                getApplicationContext().getSharedPreferences(AppConstants.PREFS_NAME, Context.MODE_PRIVATE)
                        .edit().putString(AppConstants.WEATHER_CACHE_KEY, json).apply();

                // Trigger widget update explicitly
                android.content.Intent intent = new android.content.Intent(getApplicationContext(), DayAheadWidget.class);
                intent.setAction(android.appwidget.AppWidgetManager.ACTION_APPWIDGET_UPDATE);
                int[] ids = android.appwidget.AppWidgetManager.getInstance(getApplicationContext())
                        .getAppWidgetIds(new android.content.ComponentName(getApplicationContext(), DayAheadWidget.class));
                intent.putExtra(android.appwidget.AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
                getApplicationContext().sendBroadcast(intent);

                return Result.success();
            }
        } catch (IOException e) {
            Log.e(TAG, "Widget update failed", e);
            return Result.retry();
        }

        return Result.failure();
    }
}

File: app/src/main/java/com/example/theloop/WidgetUpdateWorker.kt
package com.example.theloop

import android.content.Context
import androidx.hilt.work.HiltWorker
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.example.theloop.data.repository.CalendarRepository
import com.example.theloop.data.repository.NewsRepository
import com.example.theloop.data.repository.UserPreferencesRepository
import com.example.theloop.data.repository.WeatherRepository
import com.example.theloop.utils.AppConstants
import com.example.theloop.utils.SummaryUtils
import com.example.theloop.DayAheadWidget
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Intent
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import kotlinx.coroutines.flow.first
import java.util.ArrayList

@HiltWorker
class WidgetUpdateWorker @AssistedInject constructor(
    @Assisted appContext: Context,
    @Assisted workerParams: WorkerParameters,
    private val weatherRepo: WeatherRepository,
    private val newsRepo: NewsRepository,
    private val calendarRepo: CalendarRepository,
    private val userPreferencesRepository: UserPreferencesRepository
) : CoroutineWorker(appContext, workerParams) {

    override suspend fun doWork(): Result {
        if (!userPreferencesRepository.hasLocation()) {
            return Result.success()
        }

        return try {
            val (lat, lon) = userPreferencesRepository.location.first()
            val unit = userPreferencesRepository.tempUnit.first()
            val userName = userPreferencesRepository.userName.first()

            // Update Data (saves to DB)
            val weatherSuccess = weatherRepo.refresh(lat, lon, unit)
            val newsSuccess = newsRepo.refreshNews()
            val calendarSuccess = calendarRepo.refreshEvents()

            // Get latest data for summary
            val weather = weatherRepo.weatherData.first()
            val news = newsRepo.getArticles("US").first()
            val events = calendarRepo.events.first()

            // Generate Summary
            val summary = if (weather != null) {
                 SummaryUtils.generateSummary(
                    applicationContext,
                    weather,
                    events,
                    events.size,
                    news.firstOrNull(),
                    userName,
                    false
                )
            } else null

            if (summary != null) {
                userPreferencesRepository.saveSummary(summary)
            }

            // Trigger widget update
            val intent = Intent(applicationContext, DayAheadWidget::class.java)
            intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE)
            val ids = AppWidgetManager.getInstance(applicationContext)
                .getAppWidgetIds(ComponentName(applicationContext, DayAheadWidget::class.java))
            intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
            applicationContext.sendBroadcast(intent)

            if (!weatherSuccess || !newsSuccess || !calendarSuccess) {
                Result.retry()
            } else {
                Result.success()
            }
        } catch (e: Exception) {
            android.util.Log.e("WidgetUpdateWorker", "Widget update failed", e)
            Result.retry()
        }
    }
}
